classes/Customer.php

```diff
@@ -1193,61 +1207,80 @@ public function transformToCustomer($idLang, $password = null)
             return false;
         }
 
-        $language = new Language($idLang);
-        if (!Validate::isLoadedObject($language)) {
-            $language = Context::getContext()->language;
-        }
-
         /** @var \PrestaShop\PrestaShop\Core\Crypto\Hashing $crypto */
         $crypto = ServiceLocator::get('\\PrestaShop\\PrestaShop\\Core\\Crypto\\Hashing');
-        $this->is_guest = false;
         $this->passwd = $crypto->hash($password);
+
+        /*
+        * Now, we need to update his group. The guest should have had a PS_GUEST_GROUP previously, but if
+        * not, no biggie, it's gonna be fixed now.
+        *
+        * We will remove all entries from customer_group table and add a customer group from configuration.
+        * We also need to set it as his default group.
+        */
         $this->cleanGroups();
         $this->addGroups([Configuration::get('PS_CUSTOMER_GROUP')]);
         $this->id_default_group = (int) Configuration::get('PS_CUSTOMER_GROUP');
         $this->stampResetPasswordToken();
-        if ($this->update()) {
-            $vars = [
-                '{firstname}' => $this->firstname,
-                '{lastname}' => $this->lastname,
-                '{email}' => $this->email,
-                '{url}' => Context::getContext()->link->getPageLink(
-                    'password',
-                    true,
-                    null,
-                    sprintf(
-                        'token=%s&id_customer=%s&reset_token=%s',
-                        $this->secure_key,
-                        (int) $this->id,
-                        $this->reset_password_token
-                    )
-                ),
-            ];
-            Mail::Send(
-                (int) $idLang,
-                'guest_to_customer',
-                Context::getContext()->getTranslator()->trans(
-                    'Your guest account has been transformed into a customer account',
-                    [],
-                    'Emails.Subject',
-                    $language->locale
-                ),
-                $vars,
-                $this->email,
-                $this->firstname . ' ' . $this->lastname,
-                null,
-                null,
-                null,
-                null,
-                _PS_MAIL_DIR_,
-                false,
-                (int) $this->id_shop
-            );
 
-            return true;
+        if (!$this->update()) {
+            return false;
         }
 
-        return false;
+        $language = new Language($idLang);
```

Comment:
Why did you move this block further?

Reply:
Because it's only used for sending the email. So
A) For it to be closer to the logic where it's used.
B) It's after the early return, so it's processed only when needed.